common/src/Microsoft.Azure.IIoT.Serializers.NewtonSoft/src/Default/NewtonSoftJsonSerializer.cs

```diff
@@ -231,8 +231,17 @@ internal class JsonVariantValue : VariantValue {
             /// <inheritdoc/>
             public override object ConvertTo(Type type) {
                 try {
-                    return Token.ToObject(type,
-                        JsonSerializer.CreateDefault(_serializer.Settings));
+                        if (type != typeof(List<string>)) {
+                            return Token.ToObject(type,
+                                JsonSerializer.CreateDefault(_serializer.Settings));
+                        }
+                        else {
+                            List<string> output = new List<string>();
+                            foreach(var item in Token) {
```

Comment:
In the original code this part was never hit due to the missing DiscoveryUrl property in DiscoverySettingsController.cs.

With the other changes this part is now executed when setting a discoveryUrls, but the original code raises an exception:
  Cannot deserialize the current JSON object (e.g. {"name":"value"}) into type ' 
  System.Collections.Generic.List`1[System.String]' because the type requires a JSON array (e.g. [1,2,3]) to deserialize 
  correctly.

The reason is that Token content is something like that: {{"0": "opc.tcp://dacoldell5520:51510/UA/DemoServer"}} and doesn't correspond to type.

Reply:
The fix must then be different, because IoT Hub twin does not support array we encode all arrays in a dictionary with index as key.   To make this work make DiscoveryUrl setting on controller a Dictionary<string, string>.  Search and you will find other settings that are doing the same.   There is a helper on the dictionary called something like DecodeAsList(), use that to set the property on the discovery service.

Reply:
changed DiscoveryUlrs to Dictionary